### DISCUSSION TITLE
How to "install" a plugin?

### DISCUSSION THREAD

**Discussion by simkimsia** [kedro-org/kedro#1264] [Idea]
I'm looking at this documentation https://kedro.readthedocs.io/en/latest/07_extend_kedro/04_plugins.html#example-of-a-simple-plugin

which talks about how to write own kedro plugin but it doesn't really say about how to "install" a plugin as a user.

Maybe i overlook something.

can help?

**Comment by WaylonWalker** [comment]
Its different depending on pluign and kedro version.  It is possible for the author to design a plugin that simply works by pip installing it, but this might be dangerous for some plugins.  Others need registered.  Originally this was done by appending to the list of hooked in `ProjectContext` defined in `run.py`. I think some versions of kedro define them in `hooks.py` and the latest version registers them in `settings.py`.  Let me know if this gets you started in the right direction.

You will see this inside of `run.py` in older versions.

``` python
class ProjectContext(KedroContext):
    """Users can override the remaining methods from the parent class here,
    or create new ones (e.g. as required by plugins)
    """

    project_name = "kedro-hooks"
    # `project_version` is the version of kedro used to generate the project
    project_version = "0.15.9"
    package_name = "kedro-hooks"

    hooks = [ debug_hook ] # 👈 This is where you implement the hook

    def _get_pipelines(self) -> Dict[str, Pipeline]:
        return create_pipelines()
```

You will see something like this in your settings.py in newer versions.

``` python
"""Project settings."""
from spaceflights.hooks import ProjectHooks

# Instantiate and list your project hooks here
HOOKS = (ProjectHooks(),)

```


In all cases you will need to import the hook and add it to a tuple or list.

**Reply by datajoely** [reply]
One of the easiest things you can do is look at how `kedro-telemetry` works, it simply needs to be installed on the same environment as your main kedro project (and consent) much like @WaylonWalker describes:
https://github.com/kedro-org/kedro-plugins/tree/main/kedro-telemetry

**Reply by simkimsia** [reply]
Thanks for reply. i'm a bit confused by the marked answer. 

So do i simply use a pip install ?

Also i am curious if plugin is the way to go to have nested commands.

By that i mean  like ```kedro <plugin-name> <command-name>```

so a different plugin that uses the same command name there's no way the wrong command gets called.

Am i right?

**Reply by datajoely** [reply]
Yes so that's exactly how Kedro Viz works !
https://github.com/kedro-org/kedro-viz/blob/477db81ec6c1b58131752bd528660a290f86d3b0/package/kedro_viz/launchers/cli.py#L75

**Reply by datajoely** [reply]
By nested commands I think you will need to look into Click and the [groups feature](https://click.palletsprojects.com/en/8.0.x/api/#click.Group).

**Reply by simkimsia** [reply]
i read the groups feature of click before. So there's no issue to write a plugin such that we can have commands in the form of  `kedro <plugin-name> <command-name>` using click's groups feature yes?

**Reply by datajoely** [reply]
I can't think of a reason why it wouldn't work
